src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/CalculateStakeRule.cs

```diff
@@ -0,0 +1,44 @@
+﻿using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Consensus.Rules.CommonRules
+{
+    /// <summary>
+    /// Checks if <see cref="Block"/> has a valid PoS header.
+    /// </summary>
+    public class CalculateStakeRule : ConsensusRule
+    {
+        /// <summary>Quick easy access to the POS rules.</summary>
+        private PosConsensusRules posParent;
+
+        public override void Initialize()
+        {
+            this.posParent = this.Parent as PosConsensusRules;
+
+            Guard.NotNull(this.posParent, nameof(this.posParent));
+        }
+
+        /// <inheritdoc />
+        /// <exception cref="ConsensusErrors.HighHash"> Thrown if block doesn't have a valid PoW header.</exception>
+        public override Task RunAsync(RuleContext context)
+        {
+            context.SetStake();
+
+            if (context.Stake.BlockStake.IsProofOfWork())
+            {
+                if (context.CheckPow && !context.BlockValidationContext.Block.Header.CheckProofOfWork(context.Consensus))
+                {
+                    this.Logger.LogTrace("(-)[HIGH_HASH]");
```

Comment:
Where's the entry `()` log?

Reply:
I did not put any log entries in rules I think this can be done form the caller (I just did not add it to the caller yet).